### PR TITLE
Refactor8 mathjax v3

### PR DESCRIPTION
**Pull request by fneumann**
See  [README](Services/MathJax/README.md)

@alex40724: I moved the MathJax settings page from ilObjExternalToolsSettingsGUI to the MathJax service.
@klees: I added new setup settings and documented them in the README of the setup.

### CODE REVIEW

File: lang/ilias_en.lang [hunk 1]
```diff
@@ -5903,7 +5906,8 @@ mathjax#:#mathjax_server_cache_cleared#:#The MathJax cache was cleared.
 mathjax#:#mathjax_server_cache_size#:#Cache Size
 mathjax#:#mathjax_server_cache_size_info#:#Disk space used for caching rendered graphics.
 mathjax#:#mathjax_test_expression#:#Test Expression
-mathjax#:#mathjax_test_expression_info#:#This LaTeX expression will be rendered with the saved MathJax settings.
+mathjax#:#mathjax_test_expression_info_client#:#This expression will be rendered by tghe script in the browser.
```
Comment (mjansenDatabay) [inline lang/ilias_en.lang:5909]:
```suggestion
mathjax#:#mathjax_test_expression_info_client#:#This expression will be rendered by the script in the browser.
```

File: Services/MathJax/test/ilServicesMathJaxSuite.php [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+use PHPUnit\Framework\TestSuite;
+use PHPUnit\Framework\TestCase;
+
+/**
+ * Test Suite for the MathJax service
+ */
+class ilServicesMathJaxSuite extends TestSuite
+{
+    const TEST_CLASSES = [
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilServicesMathJaxSuite.php]:
```suggestion
    private const TEST_CLASSES = [
```

File: Services/MathJax/test/ilMathJaxBaseTest.php [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+            ->onlyMethods(['exists', 'read', 'write', 'absolutePath', 'getCacheSize'])
+            ->getMock();
+        $image->method('exists')->willReturn(false);
+        $image->method('read')->willReturn(file_get_contents(__DIR__ . '/'. $imagefile));
+        $image->method('absolutePath')->willReturn(__DIR__ . '/'. $imagefile);
+        $image->method('getCacheSize')->willReturn('10 KB');
+        return $image;
+    }
+
+    protected function getServerMock()
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilMathJaxBaseTest.php]: This method is missing the return type declaration.

File: Services/MathJax/test/ilMathJaxBaseTest.php [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+            ->getMock();
+        $factory->method('template')->willReturn($this->getTemplateMock());
+        $factory->method('server')->willReturn($this->getServerMock());
+        if (isset($imagefile)) {
+            $factory->method('image')->willReturn($this->getImageMock($imagefile));
+        }
+        return $factory;
+    }
+
+    protected function getTemplateMock()
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilMathJaxBaseTest.php]: This method is missing the return type declaration.

File: Services/MathJax/test/ilMathJaxBaseTest.php [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            false,
+            '',
+            0,
+            false,
+            false,
+            false
+        );
+    }
+
+    protected function getFactoryMock($imagefile = null)
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilMathJaxBaseTest.php]: This method is missing the typehint for the argument(s) and the return type declaration.

File: Services/MathJax/test/class.ilMathJaxTest.php [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+                       ->withServerForExport($purpose == 'export')
+                       ->withServerForPdf($purpose == 'pdf');
+
+        $mathjax = ilMathJax::getIndependent($config, $this->getFactoryMock($imagefile))->init($purpose);
+        $result = $mathjax->insertLatexImages($input);
+        $head = substr($result, 0, 60);
+        $this->assertEquals($expected, $head, 'purpose: ' . $purpose);
+    }
+
+    public function serverSideData()
```
Comment (mjansenDatabay) [inline Services/MathJax/test/class.ilMathJaxTest.php]: This method is missing the return type declaration.

File: Services/MathJax/test/class.ilMathJaxTest.php [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+            // whole expression ignored if no end is found
+            [0, '[tex]e=m*c^2[/tex][tex]e=m*c^2', null, null, '\(e=m*c^2\)[tex]e=m*c^2'],
+        ];
+    }
+
+    /**
+     * @depends testInstanceCanBeCreated
+     * @dataProvider serverSideData
+     */
+    public function testServerSideRendering($purpose, $imagefile, $expected)
```
Comment (mjansenDatabay) [inline Services/MathJax/test/class.ilMathJaxTest.php]: This method is missing the typehint for the argument(s) and the return type declaration.

File: Services/MathJax/test/class.ilMathJaxTest.php [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+        $config = $this->getEmptyConfig();
+        $mathjax = ilMathJax::getIndependent($this->getEmptyConfig(), $this->getFactoryMock());
+        $this->assertInstanceOf('ilMathJax', $mathjax);
+    }
+
+    /**
+     * @depends testInstanceCanBeCreated
+     * @dataProvider clientSideData
+     */
+    public function testClientSideRendering($limiter, $input, $start, $end, $expected)
```
Comment (mjansenDatabay) [inline Services/MathJax/test/class.ilMathJaxTest.php]: This method is missing the typehint for the argument(s) and the return type declaration.

File: Services/MathJax/test/class.ilMathJaxTest.php [hunk 9]
```diff
@@ -0,0 +1,94 @@
...
+ */
+
+require_once __DIR__ . '/ilMathJaxBaseTest.php';
+
+/**
+ * Testing the MathJax class
+ */
+class ilMathJaxTest extends ilMathJaxBaseTest
+{
+    public function testInstanceCanBeCreated()
```
Comment (mjansenDatabay) [inline Services/MathJax/test/class.ilMathJaxTest.php]: This method is missing the return type declaration.

File: Services/MathJax/interfaces/interface.ilMathJaxConfigRespository.php [hunk 10]
```diff
@@ -0,0 +1,29 @@
...
+     * Get the MathJax Configuration
+     * @return ilMathJaxConfig
+     */
+    public function getConfig() : ilMathJaxConfig;
+
+    /**
+     * Update the MathJax Configuration
+     * @param ilMathJaxConfig $config
+     */
+    public function updateConfig(ilMathJaxConfig $config);
```
Comment (mjansenDatabay) [inline Services/MathJax/interfaces/interface.ilMathJaxConfigRespository.php]: This method is missing the return type declaration.

File: Services/MathJax/classes/class.ilMathJaxSettingsGUI.php [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+    {
+        ilMathJax::getInstance()->clearCache();
+        ilUtil::sendSuccess($this->lng->txt('mathjax_server_cache_cleared'), true);
+        $this->ctrl->redirect($this);
+    }
+
+    /**
+     * Render an html link
+     */
+    protected function renderLink(string $langvar, string $url, $new_tab = true): string
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxSettingsGUI.php]:
```suggestion
    protected function renderLink(string $langvar, string $url, bool $new_tab = true): string
```

File: Services/MathJax/classes/class.ilMathJaxSettingsGUI.php [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        }
+
+        // form is not posted or has valisation errors
+        $this->tpl->setContent($this->dic->ui()->renderer()->render($form));
+    }
+
+    /**
+     * Clear the directory with cached LaTeX graphics
+     */
+    protected function clearCache()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxSettingsGUI.php]:
```suggestion
    protected function clearCache() : void
```

File: Services/MathJax/classes/class.ilMathJaxSettingsGUI.php [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+
+            default:
+                $this->tpl->setContent('unknown command: ' . $cmd);
+        }
+    }
+
+    /**
+     * Edit the MathJax settings
+     */
+    protected function editSettings()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxSettingsGUI.php]:
```suggestion
    protected function editSettings() : void
```

File: Services/MathJax/classes/class.ilMathJaxSettingsGUI.php [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+ * GUI for MathJax Settings
+ * This GUI maintains the MathJax config stored in the ILIAS settings
+ * Since ILIAS 8 these settings can also be written by the ILIAS setup
+ *
+ * @ilCtrl_Calls ilMathJaxSettingsGUI:
+ */
+class ilMathJaxSettingsGUI
+{
+    /** @var \ILIAS\DI\Container */
+    protected $dic;
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxSettingsGUI.php]: We should add typed properties for all member variables (if possible).

File: Services/MathJax/classes/class.ilMathJaxSettingsGUI.php [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        $this->repository = new ilMathJaxConfigSettingsRepository($factory);
+    }
+
+
+    /**
+     * Execute a command
+     * This should be overridden in the child classes
+     * note: permissions are already checked in the object gui
+     */
+    public function executeCommand()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxSettingsGUI.php]:
```suggestion
    public function executeCommand() : void
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 16]
```diff
@@ -0,0 +1,168 @@
...
+            $counter++;
+        }
+
+        return (round($size, 1) . " " . $type[$counter] . "B");
+    }
+
+    /**
+     * Delete all files from the cache
+     */
+    public function clearCache()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function clearCache() : void
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 17]
```diff
@@ -0,0 +1,168 @@
...
+    public function write(string $a_content)
+    {
+        $this->fs()->put($this->filepath(), $a_content);
+    }
+    
+
+    /**
+     * Get the total size of the cache with an appropriate unit for display
+     */
+    public function getCacheSize(): string
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function getCacheSize() : string
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 18]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        return $this->fs()->read($this->filepath());
+    }
+    
+    
+    /**
+     * Save the content of a cached image
+     * @param string $a_content image content
+     */
+    public function write(string $a_content)
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function write(string $a_content) : void
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 19]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    public function exists(): bool
+    {
+        return $this->fs()->has($this->filepath());
+    }
+
+    /**
+     * Read the content of a cached image
+     */
+    public function read(): string
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function read() : string
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 20]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    public function absolutePath()
+    {
+        return CLIENT_WEB_DIR . $this->filepath();
+    }
+
+    /**
+     * Check if an image is cached
+     */
+    public function exists(): bool
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function exists() : bool
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 21]
```diff
@@ -0,0 +1,168 @@
...
+        return $this->basepath
+            . '/' . substr($hash, 0, 4)
+            . '/' . substr($hash, 4, 4)
+            . '/' . $hash . $this->suffix;
+    }
+
+    /**
+     * Get the absolute path of the image
+     */
+    public function absolutePath()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function absolutePath() : string
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 22]
```diff
@@ -0,0 +1,168 @@
...
+     */
+    protected function fs()
+    {
+        return $this->di()->filesystem()->web();
+    }
+
+    /**
+     * Create the relative file path of the image
+     */
+    public function filepath()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    public function filepath() : string
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 23]
```diff
@@ -0,0 +1,168 @@
...
+    {
+        global $DIC;
+        return $DIC;
+    }
+
+    /**
+     * Get the webspace file system
+     * @return \ILIAS\Filesystem\Filesystem
+     */
+    protected function fs()
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    protected function fs() : \ILIAS\Filesystem\Filesystem
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 24]
```diff
@@ -0,0 +1,168 @@
...
+use ILIAS\DI\Container;
+
+/**
+ * Rendered MathJax image
+ * Supports image types SVG or PNG
+ * Files are stored in the web file system of ilias
+ */
+class ilMathJaxImage
+{
+    const TYPE_PNG = 'png';
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    private const TYPE_PNG = 'png';
```

File: Services/MathJax/classes/class.ilMathJaxImage.php [hunk 25]
```diff
@@ -0,0 +1,168 @@
...
+    const TYPE_SVG = 'svg';
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxImage.php]:
```suggestion
    private const TYPE_SVG = 'svg';
```

File: Services/MathJax/classes/class.ilMathJaxConfigSettingsRepository.php [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+            (bool) $this->settings->get('server_for_export'),
+            (bool) $this->settings->get('server_for_pdf')
+        );
+    }
+
+    /**
+     * Update the MathNax Configuration
+     * @param ilMathJaxConfig $config
+     */
+    public function updateConfig(ilMathJaxConfig $config)
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxConfigSettingsRepository.php]:
```suggestion
    public function updateConfig(ilMathJaxConfig $config) : void
```

File: Services/MathJax/classes/class.ilMathJaxConfigSettingsRepository.php [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+ */
+
+/**
+ * Repository for storing and loading the MathJax configiration
+ */
+class ilMathJaxConfigSettingsRepository implements ilMathJaxConfigRespository
+{
+
+    /** @var ilSetting */
+    protected $settings;
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxConfigSettingsRepository.php]:
```suggestion
    protected ilSetting $settings;
```

File: Services/MathJax/classes/class.ilMathJaxConfigSettingsRepository.php [hunk 28]
```diff
@@ -0,0 +1,71 @@
...
+ * https://github.com/ILIAS-eLearning
+ */
+
+/**
+ * Repository for storing and loading the MathJax configiration
+ */
+class ilMathJaxConfigSettingsRepository implements ilMathJaxConfigRespository
+{
+
+    /** @var ilSetting */
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxConfigSettingsRepository.php]:
```suggestion
```

File: Services/MathJax/classes/class.ilMathJaxConfig.php [hunk 29]
```diff
@@ -0,0 +1,289 @@
...
+ * https://www.ilias.de
+ * https://github.com/ILIAS-eLearning
+ */
+
+/**
+ * Global Mathjax configuration
+ */
+class ilMathJaxConfig
+{
+    const LIMITER_MATHJAX = 0;
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxConfig.php]:
```suggestion
    private const LIMITER_MATHJAX = 0;
```

File: Services/MathJax/classes/class.ilMathJaxConfig.php [hunk 30]
```diff
@@ -0,0 +1,289 @@
...
+    const LIMITER_TEX = 1;
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxConfig.php]:
```suggestion
    private const LIMITER_TEX = 1;
```

File: Services/MathJax/classes/class.ilMathJaxConfig.php [hunk 31]
```diff
@@ -0,0 +1,289 @@
...
+    const LIMITER_SPAN = 2;
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxConfig.php]:
```suggestion
    private const LIMITER_SPAN = 2;
```

File: Services/MathJax/classes/class.ilMathJaxServer.php [hunk 32]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+
+    /**
+     * Call the mathjax server
+     * @param array $options will be sent json encoded to the server
+     * @return string
+     * @throws ilMathJaxException
+     */
+    public function call(array $options): string
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxServer.php]:
```suggestion
    public function call(array $options) : string
```

File: Services/MathJax/classes/class.ilMathJaxServer.php [hunk 33]
```diff
@@ -0,0 +1,104 @@
...
+        }
+        else {
+            return $this->callByStreamContext($options);
+        }
+    }
+
+    /**
+     * Call the mathjax server by curl
+     */
+    protected function callByCurl(array $options): string
```
Comment (mjansenDatabay) [inline Services/MathJax/classes/class.ilMathJaxServer.php]:
```suggestion
    protected function callByCurl(array $options) : string
```

File: Services/MathJax/test/ilMathJaxBaseTest.php [hunk 34]
```diff
@@ -0,0 +1,88 @@
...
+    {
+        $template = $this
+            ->getMockBuilder(ilGlobalTemplate::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['addJavaScript'])
+            ->getMock();
+        return $template;
+    }
+
+    protected function getImageMock($imagefile)
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilMathJaxBaseTest.php]: This method is missing the typehint for the argument(s) and the return type declaration.

File: Services/MathJax/test/ilMathJaxBaseTest.php [hunk 35]
```diff
@@ -0,0 +1,88 @@
+<?php
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilMathJaxBaseTest.php]:
```suggestion
<?php declare(strict_types=1);
```

File: Services/MathJax/test/class.ilMathJaxTest.php [hunk 36]
```diff
@@ -0,0 +1,94 @@
...
+     */
+    public function testClientSideRendering($limiter, $input, $start, $end, $expected)
+    {
+        $config = $this->getEmptyConfig()->withClientEnabled(true)->withClientLimiter($limiter);
+        $mathjax = ilMathJax::getIndependent($config, $this->getFactoryMock());
+        $result = $mathjax->insertLatexImages($input, $start, $end);
+        $this->assertEquals($expected, $result, 'input: ' . $input);
+    }
+
+    public function clientSideData()
```
Comment (mjansenDatabay) [inline Services/MathJax/test/class.ilMathJaxTest.php]:
```suggestion
    public function clientSideData() : array
```

**Comment by fneumann** [comment]
@alex40724 @klees OK for the changes in Setup and Administration? Then I would merge the request.

File: Services/MathJax/interfaces/interface.ilMathJaxConfigRespository.php [hunk 37]
```diff
@@ -0,0 +1,26 @@
...
+    /**
+     * Get the MathJax Configuration
+     */
+    public function getConfig() : ilMathJaxConfig;
+
+    /**
+     * Update the MathJax Configuration
+     */
+    public function updateConfig(ilMathJaxConfig $config) : void;
+}
```
Comment (mjansenDatabay) [inline Services/MathJax/interfaces/interface.ilMathJaxConfigRespository.php]: An empty line is missing here. I suggest to apply the `PSR-2 + X` coding style by executing the `PHP-CS-FIXER` for all changed PHP files.

File: Services/MathJax/test/ilMathJaxBaseTest.php [hunk 38]
```diff
@@ -0,0 +1,104 @@
...
+    {
+        $server = $this
+            ->getMockBuilder(ilMathJaxServer::class)
+            ->disableOriginalConstructor()
+            ->onlyMethods(['call'])
+            ->getMock();
+        $server->method('call')->willReturn('server call result');
+        return $server;
+    }
+}
```
Comment (mjansenDatabay) [inline Services/MathJax/test/ilMathJaxBaseTest.php]: An empty line is missing here as well.

**Review by klees — APPROVED** [review]
Hi @fneumann,

changes in `./setup` LGTM.

I only skimmed over the rest of the setup related code in the component itself. Is there anything specific I could help with or that you want to ask?

Best regards!

File: Services/MathJax/interfaces/interface.ilMathJaxConfigRespository.php (hunk 37 above)
Comment (fneumann) [reply Services/MathJax/interfaces/interface.ilMathJaxConfigRespository.php]: OK, docs/development/code-style-configs/php-storm.xml seems to differ a bit from CI/PHP-CS-Fixer/code-format.php_cs regarding the newlines. Now I executed the fixer on all files.

**Comment by fneumann** [comment]
Hi @klees, thanks for your approve. The setup related code in the service is rather straightforard and I checked that it works fine.

**Pull request merged**